### PR TITLE
Thrift

### PR DESCRIPTION
**Pull request by jwolski**
Updated version of Thrift integration.

### CODE REVIEW

File: ringpop.thrift [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    1: optional list<Change> changes;
+    2: optional double membershipChecksum;
+}
+
+struct DenyingJoinsError {
+    1: required string type;
+    2: required string message;
+}
+
+struct InvalidJoinAppError {
```
Comment (kriskowal) [inline ringpop.thrift]: s/struct/exception/. It is an alias for struct for the time being.

File: server/index.js [hunk 2]
```diff
@@ -17,9 +17,13 @@
 // LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 // OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 // THE SOFTWARE.
-
 'use strict';
 
+var fs = require('fs');
+var path = require('path');
+
+var TChannelAsThrift = require('tchannel/as/thrift');
```
Comment (kriskowal) [inline server/index.js]: It is now possible to extract the .AsThrift constructor from a channel instance.

File: package.json [hunk 3]
```diff
@@ -29,6 +29,7 @@
     "farmhash": "^0.2.0",
     "metrics": "^0.1.8",
     "readable-stream": "^1.0.33",
+    "thriftify": "^1.0.0-alpha14",
```
Comment (Raynos) [inline package.json]: This should not be needed.

File: server/index.js (hunk 2 above)
Comment (Raynos) [inline server/index.js]: tchannel is not a prod dep. please use `channel.AsThrift`

File: server/index.js [hunk 4]
```diff
@@ -42,25 +46,150 @@ var commands = {
...
+    Object.keys(thriftHandlers).forEach(function eachDef(def) {
+        tchannelAsThrift.register(tchannel, def, null,
+            thriftHandlers[def]);
+    });
+}
+
+function thriftCallback(callback) {
+    return function onCallback(err, res) {
+        if (err) {
+            callback(null, new ThriftError(err));
```
Comment (Raynos) [inline server/index.js]: this is not a valid thing to do. Please check `err.nameAsThrift` before creating a ThriftError.

File: ringpop.thrift (hunk 1 above)
Comment (jwolski) [inline ringpop.thrift]: thx

File: server/index.js (hunk 2 above)
Comment (jwolski) [inline server/index.js]: good catch w.r.t the prod dep. will use AsThrift.

File: package.json (hunk 3 above)
Comment (jwolski) [inline package.json]: thx

File: server/index.js (hunk 4 above)
Comment (jwolski) [inline server/index.js]: got it. will fix.

**Pull request closed without merging**